### PR TITLE
Python 3.8+ port

### PR DESCRIPTION
**Pull request by ENDERZOMBI102**
Implement FlexVer in Python via a `FlexVer` class which implements total ordering and a `compare(str, str): int` function which is like the Java `FlexVerComparator.compare()` implementation.

This was tested on Python 3.8 up to 3.11, using the files in the `tests` folder as input.

If this is accepted, We will publish it to [PyPi](https://PyPi.org) under the `flexver` name.

- [x] I own the full rights to my contribution, and agree to release it under the terms of the Creative Commons Zero Public Domain Dedication

### CODE REVIEW

File: .github/workflows/test-python.yml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    name: test
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v3
+    - uses: actions/setup-python@v4
+      with:
+        python-version: ^3.8
+        cache: poetry
+    - run: pip install poetry # setup
+    - run: poetry build -n    # build
```
Comment (exaskye) [inline .github/workflows/test-python.yml:25]: This does not run in Actions due to the executable not being in the PATH.

**Review by exaskye — CHANGES_REQUESTED** [review]
Sorry, I missed this somehow.

Code looks OK at a glance — I'm not a Python dev, so, not really qualified to comment on that, but it's a much shorter impl than #11. I assume this is due to this impl being more OOP-y and based on the Java impl, while #11 seems to be based on the Rust impl and is more imperative.

The workflow needs to be fixed for this to be mergeable. Thanks for working on it!

File: python/flexver.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+	_original: typing.Final[ str ]
+
+	def __init__( self, version: str ) -> None:
+		self._original = version
+		self._components = _decompose( version )
+
+	def compare( self, other: 'FlexVer' ) -> int:
+		"""
+		Compares this `FlexVer` object with another.
+		\t
```
Comment (Infinidoge) [inline python/flexver.py:43]: Not sure what this `\t` is doing here

File: python/flexver.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+		return self._original
+
+	def __repr__( self ) -> str:
+		return f"FlexVer( '{self}' )"
+
+
+def compare( a: str, b: str ) -> int:
+	"""
+	Parse the given strings as freeform version strings, and compare them according to FlexVer.
+	\t
```
Comment (Infinidoge) [inline python/flexver.py:69]: Another `\t`

File: python/test_flexver.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+			if (pound := line.find('#')) != -1:
+				line = line[: pound ]
+
+			if line.endswith( '\n' ):
+				line = line[: -1]
+
+			if not line or line[0] == '#':
+				continue
+
+			setattr( FlexVerTestCase, f'test_{index}', createTest( index, line ) )
```
Comment (Infinidoge) [inline python/test_flexver.py:34]: No problem here, just stunned by how I never thought about doing something like this for dynamically generated test cases.

File: python/test_flexver.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+
+def loadTests( path: str ):
+	with open( path, 'r' ) as file:
+		for index, line in enumerate( file ):  # type: int, str
+			if (pound := line.find('#')) != -1:
+				line = line[: pound ]
+
+			if line.endswith( '\n' ):
+				line = line[: -1]
```
Comment (Infinidoge) [inline python/test_flexver.py]:
```suggestion
			line = line.rstrip('\n')
```

**Review by Infinidoge — CHANGES_REQUESTED** [review]
First and foremost, sorry for never actually finishing #11, kept meaning to but just never got around to it.

With that said, this is a great implementation, and definitely a lot cleaner than mine!
The main nitpicks I have are style, namely:
- Extra spaces inside parenthesis/brackets (which is sometimes inconsistent, such as line 120)
- `camelCase` method names instead of `snake_case` method names

These don't affect how the code runs though, so it isn't necessary to change them.

There were also some `\t`s in doc strings, and a slight code change, which I made comments for.

Thanks for picking up my dropped ball, and looking forward to seeing this on PyPI!

File: python/flexver.py (hunk 2 above)
Comment (ENDERZOMBI102) [reply python/flexver.py:43]: those are for IDEs, we've come to understand that some show reStructuredText-formatted parameters anyway, but the `\t` stops it from happening

**Comment by ENDERZOMBI102** [comment]
for the spaces between parentheses, its for readability, they help us understand divide symbols and words, make things easier to parse.
although can remove if needed.

the `camelCase` is just personal preference, will change the casing as we've also noticed that the PEP says to use `snake_case`

File: python/flexver.py (hunk 2 above)
Comment (Infinidoge) [reply python/flexver.py:43]: I do not understand IDEs sometimes. I just want my plain text left alone :(

Makes sense to have it then.

**Comment by ENDERZOMBI102** [comment]
moved to `snake_case` and fixed workflow/tests

File: python/flexver.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+		return 0 if other is _null else -other.compare_to( self )
+
+	def __str__( self ) -> str:
+		return '/'
+
+	def __repr__( self ) -> str:
+		return '/'
+
+
+_null: typing.Final[ _VersionComponent] = _NullVersionComponent()
```
Comment (Infinidoge) [inline python/flexver.py]:
```suggestion
_null: typing.Final[ _VersionComponent ] = _NullVersionComponent()
```

File: python/flexver.py [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+def compare( a: str, b: str ) -> int:
+	"""
+	Parse the given strings as freeform version strings, and compare them according to FlexVer.
+	\t
+	:param a: The first version string.
+	:param b: The second version string.
+	:returns: `0`:code: if the two versions are equal, a negative number if `a < b`:code:, or a positive number if `a > b`:code:.
+	"""
+	ad: typing.List[ _VersionComponent] = _decompose(a)
+	bd: typing.List[ _VersionComponent] = _decompose(b)
```
Comment (Infinidoge) [inline python/flexver.py]:
```suggestion
	ad: typing.List[ _VersionComponent ] = _decompose(a)
	bd: typing.List[ _VersionComponent ] = _decompose(b)
```

File: python/flexver.py [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+			out.append( createComponent( lastWasNumber, ''.join( accum ) ) )
+			accum = [ ]
+			lastWasNumber = number
+		accum.append( cp )
+
+	out.append( createComponent( lastWasNumber, ''.join( accum ) ) )
+	return out
+
+
+def _get( li: typing.List[ _VersionComponent], i: int ) -> _VersionComponent:
```
Comment (Infinidoge) [inline python/flexver.py]:
```suggestion
def _get( li: typing.List[ _VersionComponent ], i: int ) -> _VersionComponent:
```

**Review by Infinidoge — DISMISSED** [review]
Small consistency nitpicks.
Can alternatively remove the spaces around `_VersionComponent`/other types.

**Review by exaskye — APPROVED** [review]
Workflow now runs and the tests pass.

**Comment by exaskye** [comment]
Thanks!

**Comment by ENDERZOMBI102** [comment]
published! as said, can be found under `flexver` on pypi: https://pypi.org/project/flexver

**Pull request merged**